File: Breadth-First Search/Binary tree/LET_no297_Serialize and Deserialize Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return ""
        
        res = []
        queue = collections.deque([root])
        while queue:
            cur = queue.popleft()
            res.append(str(cur.val) if cur else '#')
            if cur:
                queue.append(cur.left)
                queue.append(cur.right)
        
        return ' '.join(res)

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        if not data:
            return None
        
        nodes = [TreeNode(int(val)) if val != '#' else None for val in data.split()]
        
        cur, fast = 0, 1
        while fast < len(nodes):
            while not nodes[cur]:
                cur += 1
            left, right = fast, fast + 1
            nodes[cur].left = nodes[left]
            nodes[cur].right = nodes[right]
            cur, fast = cur + 1, fast + 2
        
        return nodes[0]
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: Breadth-First Search/Binary tree/LET_no297_Serialize and Deserialize Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (!root)
            return "";
        
        queue<TreeNode*> q;
        q.push(root);
        ostringstream out;

        int count_nodes = 0;
        while (!q.empty()) {
            count_nodes++;
            TreeNode* cur = q.front();
            q.pop();
            out << (cur ? to_string(cur->val) : "#") << ' ';

            if (cur) {
                q.push(cur->left);
                q.push(cur->right);
            }
        }

        return out.str();
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if (data == "")
            return nullptr;
        
        cout << data << endl;
        istringstream in(data);
        string val;
        vector<TreeNode*> nodes;

        int count = 0;
        while (in) {
            in >> val;
            if (in)
                nodes.push_back((val == "#" ? nullptr : new TreeNode(stoi(val))));
        }

        int cur = 0, fast = 1, n = nodes.size(), left, right;
        while (fast < n) {
            while (!nodes[cur])
                cur++;

            left = fast;
            right = fast + 1;
            nodes[cur]->left = nodes[left];
            nodes[cur]->right = nodes[right];
            cur++;
            fast += 2;
        }

        return nodes[0];
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec2 {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        ostringstream out;
        serialHelper(root, out);
        return out.str();
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        istringstream in(data);
        return deserialHelper(in);
    }

    void serialHelper(TreeNode* root, ostringstream& out) {
        if (root) {
            out << root->val << ' ';
            serialHelper(root->left, out);
            serialHelper(root->right, out);
        }
        else
            out << "# ";
    }

    TreeNode* deserialHelper(istringstream& in) {
        string val;
        in >> val;
        if (val == "#")
            return nullptr;

        TreeNode* cur = new TreeNode(stoi(val));
        cur->left = deserialHelper(in);
        cur->right = deserialHelper(in);
        return cur;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));